asaplib/pca/ml_kpca.py

```diff
@@ -9,6 +9,207 @@
+        self._m = len(kmat)
+        self.colmean, self.mean, self.center_kmat = self.center_square(kmat)
+
+        # calculation, ordering in descending order and scale by sqrt(lambda)
+        self._lambdas, self._alphas = salg.eigh(self.center_kmat, eigvals=(self._m - self.ndim, self._m - 1))
+        self._lambdas = np.flipud(self._lambdas)
+        self._alphas = np.fliplr(self._alphas) / np.sqrt(self._lambdas)
+
+        # the model is fitted then
+        self._fitted = True
+
+    def fit_transform(self, kmat, copy=True):
+        """Fit kernel PCA on the precomputed kernel matrix & project to lower dimensions
+
+        Notes
+        -----
+        - Keeps the kernel matrix intact
+        - can be done only once on a given object, need to reinitialise if trying to run again
+
+        Parameters
+        ----------
+        kmat : numpy.ndarray, shape=(M,M)
+            kerenl matrix between the observations
+            needs to be square and real, symmetric
+        copy : bool, optional, default=True
+            copy the kernel matrix or overwrite it, passed to self.transform()
+            nb. the kernel matrix will be left centered if this is False
+
+        Returns
+        -------
+
+        """
+        self.fit(kmat)
+        return self.transform(self.center_kmat, iscentered=True, copy=copy)
+
+    def transform(self, ktest, iscentered=False, copy=True):
+        """Transforms to the lower dimensions
+
+        Parameters
+        ----------
+        ktest : array_like, shape=(L,M)
+            kernel matrix of the test vectors with the training vectors
+        iscentered : bool, optional, default=False
+            if the kernel is centered already, mainly used for the fit_transform function
+        copy : bool, optional, default=True
+            copy the kernel matrix or overwrite it
+            nb. the kernel matrix will be left centered if this is False
+
+        Returns
+        -------
+        projected_vectors: numpy.array, shape=(L,ndim)
+            projections of the given kernel wrt. the training kernel
+        """
+        if not self._fitted:
+            raise RuntimeError("The model has not been fitted yet, please fit it and then use transform.")
+
+        self._check_kmat(ktest, square=False)
+
+        if iscentered:
+            centered_ktest = ktest
+        else:
+            centered_ktest = self._center_test_kmat((ktest.copy() if copy else ktest))
+
+        return np.dot(centered_ktest, self._alphas)
+
+    def _check_kmat(self, kmat, square=True):
+        """Check the kernel matrix
+
+        checks if the matrix is real valued and its shape: square symmetric if square=True, LxM(self._m) otherwise
+
+        Parameters
+        ----------
+        kmat : numpy.array
+            kernel matrix to be checked
+        square : bool, default=True
+            if the matrix needs to be square and symmetric
+
+        Returns
+        -------
+
+        """
+
+        assert np.ndim(kmat) == 2
```

Comment:
We're only every going to do 2D visualisations I guess?

Reply:
nay, this is asserting that the kernel matrix is a 2D array, but not affecting the number of principal components that we are calculating